src/ufoLib2/objects/glyph.py

```diff
@@ -292,33 +310,33 @@ def move(self, delta: Tuple[Number, Number]) -> None:
     # Pen methods
     # -----------
 
-    def draw(self, pen):
+    def draw(self, pen: AbstractPen) -> None:
         """Draws glyph into given pen."""
         # TODO: Document pen interface more or link to somewhere.
         pointPen = PointToSegmentPen(pen)
         self.drawPoints(pointPen)
 
-    def drawPoints(self, pointPen):
+    def drawPoints(self, pointPen: AbstractPointPen) -> None:
         """Draws points of glyph into given point pen."""
         for contour in self.contours:
             contour.drawPoints(pointPen)
         for component in self.components:
             component.drawPoints(pointPen)
 
-    def getPen(self):
+    def getPen(self) -> SegmentToPointPen:
```

Comment:
Should this return `AbstractPen` instead? By returning `SegmentToPointPen` you're sort of promising that the method will always return that, when actually that's an implementation detail. Same for the next method, `GlyphPointPen` is implementation, I would say.

Reply:
Mh, I read in a typing article somewhere that one should be generic for input types but specific for output types. If the implementation changes, we just change the return type, no?

Reply:
Ok, I would imagine an argument in favour of the strategy you describe being: don't try to abstract things away, it's much easier for users of the method to know exactly what they're getting instead of some abstraction.

However in this particular case, I would say that it's nobody's business outside of the Glyph class to know that actually the segment pen is implemented in terms of the point pen using this translation helper class. The fact that it's a `SegmentToPointPen` is at best not relevant to the caller, and at worst will allow the caller to depend on something specific that `SegmentToPointPen` has and that `AbstractPen` in general doesn't have, thus creating a bug if for example another library that can be swapped in place of ufoLib2 like defcon implements glyph pens in another way, or if ufoLib2 wants to change its glyph pen implementation later. 

(I'm blowing potential problems out of proportion here just to make my point, in practice I feel that it wouldn't matter that much for the case of this particular method).

Reply:
Regarding abstract input, concrete output: https://discuss.python.org/t/typing-recommendation-abstract-input-types-concrete-return-types/3989/2

K, then I'll change it to something abstract.